src/test/netlink/socket/bind/test_bind.rs

```diff
@@ -59,11 +59,15 @@ fn get_tests() -> Vec<test_utils::ShadowTest<(), String>> {
                     move || test_short_addr(sock_type, flag),
                     set![TestEnv::Libc, TestEnv::Shadow],
                 ),
+                // This test is tested only in Shadow because Linux supports all the groups, while
+                // Shadow does not.
                 test_utils::ShadowTest::new(
                     &append_args("test_supported_groups"),
                     move || test_supported_groups(sock_type, flag),
                     set![TestEnv::Shadow],
                 ),
+                // This test is tested only in Shadow because Linux supports all the groups, while
+                // Shadow does not.
                 test_utils::ShadowTest::new(
                     &append_args("test_unsupported_groups"),
                     move || test_unsupported_groups(sock_type, flag),
```

Comment:
I think we want to write our tests such that the behavior that Shadow does implement is tested in both Linux AND Shadow, to make sure that Shadow's behavior is consistent with Linux. For behavior that Shadow does not support, that can be tested in just Linux.

It looks like here we are writing only Shadow-specific tests, but we have no way of comparing Shadow behavior to Linux behavior. Am I missing something?

Reply:
> For behavior that Shadow does not support, that can be tested in just Linux.

`test_unsupported_groups` makes sure that, when the unsupported features are used, Shadow will emit an error.

I think testing the unsupported features only in Linux and not caring what Shadow does with it make less sense.

What do you think?

> It looks like here we are writing only Shadow-specific tests, but we have no way of comparing Shadow behavior to Linux behavior. Am I missing something?

Linux is supposed to support all the groups. Perhaps, I should do `test_supported_groups` in both Shadow and Linux, but do `test_unsupported_groups` only in Shadow.

Reply:
OK, got it. Yeah normally we want to write tests that pass in Linux, with the goal of (eventually) making them also pass in Shadow (ie so Shadow matches Linux). So we would write something like a `test_group` function to be modular and take a group, and then specify which groups should pass in linux (probably all of them) and which should pass in Shadow.

In this case you've already separated the ones that should pass in Shadow. So I agree that you should do `test_supported_groups` in both Shadow and Linux.

For `test_unsupported_groups`, you check that Shadow is returning the correct error, which is a good start. We still don't know what Linux is doing in those cases, but I guess that's OK for now since we're not planning to add support for those in Shadow yet. If we do eventually add support in Shadow, then we could add the Linux+Shadow tests at that time.